tools/pythonpkg/src/pyconnection.cpp

```diff
@@ -1530,31 +1530,62 @@ static unique_ptr<TableRef> TryReplacement(py::dict &dict, py::str &table_name,
 	return std::move(table_function);
 }
 
+static optional_idx GetLocalLimit(ClientContext &context) {
+	Value result;
+	if (!context.TryGetCurrentSetting("local_scan_frames", result)) {
+		return optional_idx();
+	}
+	auto limit = result.GetValue<uint64_t>();
+	return optional_idx(limit);
+}
+
+static optional_idx GetGlobalLimit(ClientContext &context) {
+	Value result;
+	if (!context.TryGetCurrentSetting("global_scan_frames", result)) {
+		return optional_idx();
+	}
+	auto limit = result.GetValue<uint64_t>();
+	return optional_idx(limit);
+}
+
 static unique_ptr<TableRef> ScanReplacement(ClientContext &context, const string &table_name,
                                             ReplacementScanData *data) {
 	py::gil_scoped_acquire acquire;
 	auto py_table_name = py::str(table_name);
 	// Here we do an exhaustive search on the frame lineage
 	auto current_frame = py::module::import("inspect").attr("currentframe")();
 	auto client_properties = context.GetClientProperties();
+
+	auto global_limit = GetGlobalLimit(context);
+	auto local_limit = GetLocalLimit(context);
+
+	idx_t depth = 0;
 	while (hasattr(current_frame, "f_locals")) {
 		auto local_dict = py::reinterpret_borrow<py::dict>(current_frame.attr("f_locals"));
```

Comment:
I wonder if we should search the entire stack by default. It seems this behavior was introduced in https://github.com/duckdb/duckdb/pull/3000 - but that might be have been too heavy-handed. Perhaps we should respect Pythons' variable resolution by default instead?

Since that's a breaking change we would need to be careful with making it, though.

Reply:
I see, so we should default to a depth of 1 and only through manual setting would the behavior that #3000 fixes be enabled?

Reply:
I don't think we should do it right now but we might want to move towards something like that in the future, I would say DuckDB should use the same variable resolution rules as Python itself ideally

Reply:
Feels like Python's `locals()` and `globals()` should be used  for variable resolution in this order (which is how Python resolves variables), and the only sensible solution for  #3000 seems to be to perform the replacement scans at the time of the `query` call, not at the time of query execution. That would fix a few other issues too, e.g. #11687.  If it's not possible to perform the replacement scans at the time of query, a `context` parameter to `query` (#10191), and disallowing any other replacements, would be a clean solution (which I personally prefer over automatic replacement scans anyway).

Reply:
I think you're right, when the replacement scan is done only once, at `query` call then that issue should be resolved.
That is the solution I have in mind for <https://github.com/duckdb/duckdb/issues/11687>, it's been a bit of a technical challenge though. I have a branch where that is functional, but we're not really happy with the way it's implemented so we'll have to find a better one.

Because this work has to be done during the replacement scan which is part of the Bind, this work is done in core and we have limited control. I don't think there is a way to do this without making changes in core, but never the less we would like to limit how impactful those changes are.

Reply:
> a `context` parameter to query

I am aware, and you have been vocal about that, but that is not going to happen because that would change existing behavior for the worse.
As mentioned you can get this behavior by using `.register` combined with this new setting to 0.

With the power of Python you could even patch the existing `sql` and `query` methods to decorate it with a `context` parameter that does:

```py
cursor = con.cursor()
cursor.execute("set local_scan_frames=0")
cursor.execute("set global_scan_frames=0")
context = kwargs.pop("context")
if context:
    for (key, value in context.items()):
        cursor.register(key, value)
cursor.query(*args, **kwargs)
```

Reply:
I'm not writing the `context` option off entirely, perhaps it makes sense to disable replacement scans **only** when the context parameter is provided and use that instead, but keep the default behavior the same.